x/csr/types/params.go

```diff
@@ -1,32 +1,100 @@
 package types
 
 import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
 )
 
 var _ paramtypes.ParamSet = (*Params)(nil)
 
-// ParamKeyTable the param key table for launch module
+var (
+	DefaultEnableCSR                   = false
+	DefaultCSRShares                   = sdk.NewDecWithPrec(50, 2)
+	DefaultAddressDerivationCostCreate = uint64(50)
+
+	ParamStoreKeyEnableCSR                   = []byte("EnableCSR")
+	ParamStoreKeyCSRShares                   = []byte("CSRShares")
+	ParamStoreKeyAddressDerivationCostCreate = []byte("AddressDerivationCostCreate")
+)
+
+// ParamKeyTable the param key table
 func ParamKeyTable() paramtypes.KeyTable {
 	return paramtypes.NewKeyTable().RegisterParamSet(&Params{})
 }
 
 // NewParams creates a new Params instance
-func NewParams() Params {
-	return Params{}
+func NewParams(enableCSR bool, csrShares sdk.Dec, addressDerivationCostCreate uint64) Params {
+	return Params{
+		EnableCsr:                   enableCSR,
+		CsrShares:                   csrShares,
+		AddressDerivationCostCreate: addressDerivationCostCreate,
+	}
 }
 
 // DefaultParams returns a default set of parameters
 func DefaultParams() Params {
-	return NewParams()
+	return NewParams(DefaultEnableCSR, DefaultCSRShares, DefaultAddressDerivationCostCreate)
 }
 
 // ParamSetPairs get the params.ParamSet
 func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
-	return paramtypes.ParamSetPairs{}
+	return paramtypes.ParamSetPairs{
+		paramtypes.NewParamSetPair(ParamStoreKeyEnableCSR, &p.EnableCsr, ValidateEnableCSR),
+		paramtypes.NewParamSetPair(ParamStoreKeyCSRShares, &p.CsrShares, ValidateShares),
+		paramtypes.NewParamSetPair(ParamStoreKeyAddressDerivationCostCreate, &p.AddressDerivationCostCreate, ValidateAddressDerivationCostCreate),
+	}
 }
 
-// Validate validates the set of params
-func (p Params) Validate() error {
+// Validates the cost of address derivation
+func ValidateAddressDerivationCostCreate(i interface{}) error {
+	_, ok := i.(uint64)
+	if !ok {
+		return sdkerrors.Wrapf(ErrInvalidParams, "Params::Validate::ValidateAddressDerivationCostCreate address derivation cost must be a uint.")
```

Comment:
Should we also validate that addrDerivationCost create is non-zero?

Reply:
we could make it free to the network i guess its just something that has to be decided on genesis

Reply:
i feel like its probably fine